### ISSUE TITLE
Pleats filter simulator

### ISSUE THREAD

**Issue by aizim114** [editorconfig/editorconfig-vscode#456]
HTML

<!DOCTYPE html>
<html lang="ko">
<head>
    <meta charset="UTF-8">
    <title>필터 절곡산 시뮬레이터</title>
    <script src="https://cdn.jsdelivr.net/npm/axios/dist/axios.min.js"></script>
    <script src="https://cdn.plot.ly/plotly-2.32.0.min.js"></script>
    <style>
        body {font-family: Arial, sans-serif; margin:30px;}
        .container {max-width:800px; margin:auto;}
        label {display:block; margin-top:12px;}
        input, select {width:100%; padding:6px; margin-top:4px;}
        button {margin-top:15px; padding:8px 16px;}
        #chart {width:100%; height:500px; margin-top:20px;}
        .result {margin-top:15px; font-weight:bold;}
    </style>
</head>
<body>
<div class="container">
    <h2>필터 절곡산 시뮬레이터</h2>

    <label for="material">원단 종류</label>
    <select id="material">
        <option value="MERV8">MERV8</option>
        <option value="MERV11">MERV11</option>
        <option value="MERV12">MERV12</option>
        <option value="MERV13">MERV13</option>
    </select>

    <label for="height">필터산 높이 (mm)</label>
    <input type="number" id="height" min="1" step="1" placeholder="예: 20">

    <button id="run">시뮬레이션 실행</button>

    <div id="result" class="result"></div>

    <div id="chart"></div>
</div>

<script>
/* ---------- 1) CSV 로드 (브라우저 내에서) ---------- */
async function loadCSV(url) {
    const resp = await fetch(url);
    const text = await resp.text();
    const rows = text.trim().split('\n').map(r=>r.split(','));
    const header = rows[0];
    const data = rows.slice(1).map(r=>{
        const obj = {};
        r.forEach((v,i)=> obj[header[i]] = v.trim());
        return obj;
    });
    return data;
}

/* ---------- 2) 보간·최적화 로직 ---------- */
function buildLookup(csvData){
    // {material: {height: {bend: loss}}}
    const lookup = {};
    csvData.forEach(row=>{
        const mat = row.material;
        const h = +row.height_mm;
        const b = +row.bend_per;
        const loss = +row.loss_pa;
        if(!lookup[mat]) lookup[mat] = {};
        if(!lookup[mat][h]) lookup[mat][h] = {};
        lookup[mat][h][b] = loss;
    });
    return lookup;
}

/* 선형 보간 (bend_per 기준) */
function interpolateBend(bendMap, targetBend){
    const bends = Object.keys(bendMap).map(Number).sort((a,b)=>a-b);
    const losses = bends.map(b=>bendMap[b]);
    // 1‑D 선형 보간
    const idx = bends.findIndex(b=>b>=targetBend);
    if(idx===0) return losses[0];
    if(idx===bends.length) return losses[losses.length-1];
    const x0 = bends[idx-1], y0 = losses[idx-1];
    const x1 = bends[idx],   y1 = losses[idx];
    return y0 + (y1-y0)*(targetBend-x0)/(x1-x0);
}

/* 높이 보간 (height 기준) */
function interpolateHeight(lookup, material, targetHeight, step){
    const heights = Object.keys(lookup[material]).map(Number).sort((a,b)=>a-b);
    // 목표 높이가 범위 밖이면 가장 가까운 값 반환
    if(targetHeight <= heights[0]) return interpolateBend(lookup[material][heights[0]], step);
    if(targetHeight >= heights[heights.length-1]) return interpolateBend(lookup[material][heights[heights.length-1]], step);
    // 목표 높이가 기존 높이와 정확히 일치하면 바로 반환
    if(heights.includes(targetHeight)){
        return interpolateBend(lookup[material][targetHeight], step);
    }
    // 선형 보간 (두 인접 높이 사이)
    const lower = heights.filter(h=>h<targetHeight).sort((a,b)=>b-a)[0];
    const upper = heights.filter(h=>h>targetHeight).sort((a,b)=>a-b)[0];
    const yLow = interpolateBend(lookup[material][lower], step);
    const yUp  = interpolateBend(lookup[material][upper], step);
    return yLow + (yUp-yLow)*(targetHeight-lower)/(upper-lower);
}

/* ---------- 3) 메인 시뮬레이션 ---------- */
async function runSimulation(){
    const material = document.getElementById('material').value;
    const height   = parseFloat(document.getElementById('height').value);
    if(isNaN(height) || height<=0){
        alert('필터산 높이를 올바르게 입력하세요.');
        return;
    }

    const csvData = await loadCSV('filter_data.csv');   // 같은 폴더에 있다고 가정
    const lookup  = buildLookup(csvData);

    // 1) 절곡산 수 범위(예: 40~140, step=5) 를 고정하거나 CSV에 있는 최소·최대값을 사용
    const allBends = Object.values(lookup[material])
                         .flatMap(hMap=>Object.keys(hMap).map(Number))
                         .sort((a,b)=>a-b);
    const minBend = Math.min(...allBends);
    const maxBend = Math.max(...allBends);
    const step = 5;   // 원하는 간격

    const curve = [];
    for(let b=minBend; b<=maxBend; b+=step){
        const loss = interpolateHeight(lookup, material, height, b);
        curve.push({bend:b, loss:loss});
    }

    // 최소값 찾기
    const minIdx = curve.reduce((i,cur,idx)=> (cur.loss<curve[i].loss)?idx:i,0);
    const optimal = curve[minIdx];

    // ---------- 4) 차트 그리기 ----------
    const trace = {
        x: curve.map(p=>p.bend),
        y: curve.map(p=>p.loss),
        mode:'lines+markers',
        type:'scatter',
        line:{shape:'spline'},
        marker:{size:6}
    };
    const layout = {
        title:`${material} – 필터산 높이 ${height} mm`,
        xaxis:{title:'절곡산 수 (#/500 또는 #/594)'},
        yaxis:{title:'압력손실 (Pa)'},
        hovermode:'closest'
    };
    Plotly.newPlot('chart',[trace],layout);

    // 최소점 강조
    Plotly.addTraces([{
        x:[optimal.bend],
        y:[optimal.loss],
        mode:'markers',
        marker:{color:'red',size:12,symbol:'star'},
        name:'최적 절곡산'
    }],'chart');

    // ---------- 5) 결과 텍스트 ----------
    document.getElementById('result').innerHTML =
        `<strong>최적 절곡산:</strong> ${optimal.bend}<br>
         <strong>그때 압력손실:</strong> ${optimal.loss.toFixed(2)} Pa`;
}

/* 버튼 클릭 이벤트 */
document.getElementById('run').addEventListener('click', runSimulation);
</script>
</body>
</html>
